cfdm/read_write/netcdf/netcdfread.py

```diff
@@ -9883,3 +10039,273 @@ def _ugrid_check_connectivity_variable(
+        """Return a dimension is size.
+
+        .. versionadded:: (cfdm) NEXTVERSION
+
+        :Parameters:
+
+            nc: `netCDF4.Dataset` or `h5netcdf.File`
+                The dataset.
+
+            dim_name: `str`
+                The dimension name.
+
+        :Returns:
+
+            `int`
+                The dimssion size
+
+        """
+        return self._file_dimension(nc, dim_name).size
+
+    def _file_variables(self, nc):
+        """Return all variables in the root group.
+
+        .. versionadded:: (cfdm) NEXTVERSION
+
+        :Parameters:
+
+            nc: `netCDF4.Dataset` or `h5netcdf.File`
+                The dataset.
+
+        :Returns:
+
+            `dict'-like
+                A dictionary of the variables keyed by their names.
+
+        """
+        return nc.variables
+
+    def _file_variable(self, nc, var_name):
+        """Return a variable.
+
+        .. versionadded:: (cfdm) NEXTVERSION
+
+        :Parameters:
+
+            nc: `netCDF4.Dataset` or `h5netcdf.File`
+                The dataset.
+
+            var_name: `str`
+                The variable name.
+
+        :Returns:
+
+            `netCDF4.Variable` or `h5netcdf.Variable`
+                The variable.
+
+        """
+        return self._file_variables(nc)[var_name]
+
+    def _file_variable_attributes(self, var):
+        """Return the variable attributes.
+
+        .. versionadded:: (cfdm) NEXTVERSION
+
+        :Parameters:
+
+            var: `netCDF4.Variable` or `h5netcdf.Variable`
+                The variable.
+
+        :Returns:
+
+            `dict`
+                A dictionary of the attribute values keyed by their
+                names.
+
+        """
+        try:
+            # h5netcdf
+            return dict(var.attrs)
+        except AttributeError:
+            # netCDF4
+            return {attr: var.getncattr(attr) for attr in var.ncattrs()}
+
+    def _file_variable_dimensions(self, var):
+        """Return the variable dimension names.
+
+        .. versionadded:: (cfdm) NEXTVERSION
+
+        :Parameters:
+
+            var: `netCDF4.Variable` or `h5netcdf.Variable`
+                The variable.
+
+        :Returns:
+
+            `tuple` or `str`
+                The dimension names.
+
+        """
+        return var.dimensions
+
+    def _file_variable_size(self, var):
+        """Return the size of a variable's array.
+
+        .. versionadded:: (cfdm) NEXTVERSION
+
+        :Parameters:
+
+            var: `netCDF4.Variable` or `h5netcdf.Variable`
+                The variable.
+
+        :Returns:
+
+            `int`
+                The array size.
+
+        """
+        # Use try/except here because the variable type could differ
+        # from that implied by the value of self.read_vars["netCDF4"]
+        try:
+            # netCDF4
+            return var.size
+        except AttributeError:
+            # h5netcdf
+            return prod(var.shape)
```

Comment:
I think that the APIs determine the order. In this case `netCDF4` has to come first because `h5netcdf` does not have a `size` attribute. OK, so both have a `shape` attribute (in this case), but we don't want to do work that's already been done. In some cases (e.g. `def _file_global_attribute(self, nc, attr):` the order doesn't matter.

We may introduce a _third_ backend (h4netcdf driven by pyfive) - if so we'll revisit this topic for sure.